lib/jekyll/hooks.rb

```diff
@@ -0,0 +1,103 @@
+module Jekyll
+  module Hooks
+    # Helps look up hooks from the registry by owner's class
+    OWNER_MAP = {
+      Jekyll::Site => :site,
+      Jekyll::Page => :page,
+      Jekyll::Post => :post,
+      Jekyll::Document => :document,
+    }.freeze
+
+    DEFAULT_PRIORITY = 20
+
+    # compatibility layer for octopress-hooks users
+    PRIORITY_MAP = {
+      low: 10,
+      normal: 20,
+      high: 30,
+    }.freeze
+
+    # initial empty hooks
+    @registry = {
+      :site => {
+        reset: [],
+        post_read: [],
+        pre_render: [],
+        post_write: [],
+      },
+      :page => {
+        post_init: [],
+        pre_render: [],
+        post_render: [],
+        post_write: [],
+      },
+      :post => {
+        post_init: [],
+        pre_render: [],
+        post_render: [],
+        post_write: [],
+      },
+      :document => {
+        pre_render: [],
+        post_render: [],
+        post_write: [],
+      },
+    }
+
+    # map of all hooks and their priorities
+    @hook_priority = {}
+
+    NotAvailable = Class.new(RuntimeError)
+
+    # register hook(s) to be called later
+    def self.register(owners, event, priority: nil, &block)
```

Comment:
I personally don't like this usage of keyword args.  Why not have it default the value to the internal API's instead of having nil seep through and then deciding on the outside, have the outside default and the internal API force a value or collapse.  That's just my opinion: the public API's set the usage and defaults.